hexkit/protocols/dao.py

```diff
@@ -0,0 +1,415 @@
+    the DAO. Thus, both a standard DTO model (first type variable), which includes the
+    key field, as well as special DTO model (second type variable), which is identical
+    to the does not include the ID field and is dedicated for creation of new resources,
+    is needed.
+    """
+
+    def insert(self, dto: DtoCreation_contra) -> Dto:
+        """Create a new resource.
+
+        Args:
+            dto:
+                Resource content as a pydantic-based data transfer object without the
+                resource ID (which will be set automatically).
+
+        Returns:
+            Returns a copy of the newly inserted resource including it assigned ID.
+        """
+        ...
+
+
+class DaoNaturalId(DaoCommons[Dto], typing.Protocol[Dto]):
+    """A duck type of a DAO that uses natural resource ID profided by the client."""
+
+    def insert(self, dto: Dto) -> None:
+        """Create a new resource.
+
+        Args:
+            dto:
+                Resource content as a pydantic-based data transfer object including the
+                resource ID.
+
+        Raises:
+            ResourceAlreadyExistsError:
+                when a resource with the id specified in the dto does already exist.
+        """
+        ...
+
+    def upsert(self, dto: Dto) -> None:
+        """Update the provided resource if it already exists, create it otherwise.
+
+        Args:
+            dto:
+                Resource content as a pydantic-based data transfer object including the
+                resource ID.
+        """
+        ...
+
+
+class DaoFactoryProtcol(ABC):
+    """A Protocol describing a factory to produce Data Access Objects (DAO) objects when
+    providing a Data Transfer Objetct (DTO) class.
+    """
+
+    class IdFieldNotFoundError(ValueError):
+        """Raised when the dto_model did not contain the expected id_field."""
+
+    class CreationModelInvalidError(ValueError):
+        """Raised when the DtoCreationModel was invalid in relation to the main
+        DTO model."""
+
+    class IndexFieldsInvalidError(ValueError):
+        """Raised when providing an invalid list of fields to index."""
+
+    @classmethod
+    def _validate_dto_model_id(cls, *, dto_model: type[Dto], id_field: str) -> None:
+        """Checks whether the dto_model contains the expected id_field.
+        Raises otherwises."""
+
+        if id_field not in dto_model.schema()["properties"].keys():
+            raise cls.IdFieldNotFoundError()
+
+    @classmethod
+    def _validate_dto_creation_model(
+        cls,
+        *,
+        dto_model: type[Dto],
+        dto_creation_model: Optional[type[DtoCreation]],
+        id_field: str,
+    ) -> None:
+        """Checks that the dto_creation_model has identical fields than the dto_model
+        except missing the ID. Raises otherwises."""
+
+        if dto_creation_model is None:
+            return
+
+        expected_properties = deepcopy(dto_model.schema()["properties"])
```

Comment:
simple copy should be sufficient here I think

Reply:
✅